### ISSUE TITLE
Outrageous pre-training time

### ISSUE THREAD

**Issue by WHlTE-N0lSE** [cvlab-columbia/hyperfuture#3]
How much time did you spend pretraining the model? I experimented on kinetics400, but found that the pre-training time is at least 30 days. I train with 8 GeForce RTX 2080 Tis. This training time is outrageous.
![图片1](https://user-images.githubusercontent.com/74521283/200800115-9ed57dff-dff2-4566-a8ce-8775c0a3c0ef.png)
This is my setting:
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 PYTORCH_JIT=0 NCCL_LL_THRESHOLD=0 python \
  -W ignore \
  -i \
  -m torch.distributed.launch \
  --master_port=9999 \
  --nproc_per_node=8 \
  main.py \
  --pred_step 3 \
  --hyperbolic \
  --hyperbolic_version 1 \
  --distance squared \
  --network_feature resnet18-dim8 \
  --dataset kinetics \
  --seq_len 5 \
  --num_seq 8 \
  --ds 3 \
  --batch_size 64 \
  --img_dim 112 \
  --epochs 100 \
  --fp16 \
  --fp64_hyper \
  --num_workers 8 \
  --lr 0.001 \
  --prefix train_kinetics_hyperbolic \
  --path_dataset /data/video_data/kinetics-400/kinetics-400/split/ \
  --path_data_info /data/video_data/kinetics-400/kinetics-400/split/

**Comment by ruoshiliu** [comment]
Our pretraining time is around 2-4 days on a 4-GPU machine with RTX 2080 Ti. It looks like either a bug in your code or something is severely bottlenecking your training. Did you check the average GPU utilization? Is it close to at least 80%?

**Comment by WHlTE-N0lSE** [comment]
No,it's almost 100%




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "cvlab-columbia/hyperfuture"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年11月9日(星期三) 晚上6:04
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [cvlab-columbia/hyperfuture] Outrageous pre-training time (Issue #3)





 
Our pretraining time is around 2-4 days on a 4-GPU machine with RTX 2080 Ti. It looks like either a bug in your code or something is severely bottlenecking your training. Did you check the average GPU utilization? Is it close to at least 80%?

**Comment by WHlTE-N0lSE** [comment]
No,it's almost 100%





------------------ 原始邮件 ------------------
发件人:                                                                                                                        "cvlab-columbia/hyperfuture"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年11月9日(星期三) 晚上6:04
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [cvlab-columbia/hyperfuture] Outrageous pre-training time (Issue #3)





 
Our pretraining time is around 2-4 days on a 4-GPU machine with RTX 2080 Ti. It looks like either a bug in your code or something is severely bottlenecking your training. Did you check the average GPU utilization? Is it close to at least 80%?

**Comment by WHlTE-N0lSE** [comment]
I modified the dataloader to load videos instead of images, is that the reason? If Kinetics400 is stored in the form of pictures, it will occupy a huge amount of hard disk space.
This is my code:
    def __getitem__(self, index):
        #import ipdb;ipdb.set_trace()
        vpath=str(self.video_info.iloc[index]).split()[1]
        vpath="/data/video_data/kinetics-400/kinetics-400/train/"+vpath
        vlen=int(cv2.VideoCapture(vpath).get(cv2.CAP_PROP_FRAME_COUNT))
        items = self.idx_sampler(vlen, vpath)
        while items is None or items[0] is None:
            index = random.randint(0, len(self.video_info) - 1)
            vpath, vlen = self.video_info.iloc[index]
            items = self.idx_sampler(vlen, vpath)

        idx_block, vpath = items
        assert idx_block.shape == (self.num_seq, self.seq_len)
        idx_block = idx_block.reshape(self.num_seq * self.seq_len)
        #import ipdb;ipdb.set_trace()
        video_reader = decord.VideoReader(os.path.join(self.path_dataset, vpath), num_threads=1)
        #video_reader = decord.VideoReader(os.path.join(self.path_dataset, vpath), num_threads=10)
        video_data = video_reader.get_batch(idx_block).asnumpy()
        seq=[Image.fromarray(cv2.cvtColor(frameK,cv2.COLOR_BGR2RGB)) for frameK in video_data]
        #seq = [pil_loader(os.path.join(self.path_dataset, vpath, 'image_%05d.jpg' % (i + 1))) for i in idx_block]
        t_seq = self.transform(seq)  # apply same transform
        (C, H, W) = t_seq[0].size()
        t_seq = torch.stack(t_seq, 0)
        t_seq = t_seq.view(self.num_seq, self.seq_len, C, H, W).transpose(1, 2)

        label_text = vpath.split('/')[-2]
        label = self.label_to_id[label_text]

        if self.return_idx:
            return t_seq, label, index
        return t_seq, label

**Comment by WHlTE-N0lSE** [comment]
I modified the dataloader to load videos instead of images, is that the reason? If Kinetics400 is stored in the form of pictures, it will occupy a huge amount of hard disk space.
This is my code:
video_reader = decord.VideoReader(os.path.join(self.path_dataset, vpath), num_threads=1)
video_data = video_reader.get_batch(idx_block).asnumpy()
seq=[Image.fromarray(cv2.cvtColor(frameK,cv2.COLOR_BGR2RGB)) for frameK in video_data]

**Comment by surisdi** [comment]
Hi @WHlTE-N0lSE, loading separate images instead of videos can make the loading much faster depending on the setting. In our case, it helped, but its importance depends on different factors, like the type of disk you are using or how many frames per video are required. But if you see 100% utilization of the GPU, loading time is not the issue. Can you confirm that your GPU utilization (not memory) is very high? GPU memory can stay high while the GPU is not really being utilized because it is waiting for data.

**Comment by WHlTE-N0lSE** [comment]
Thank you, this did speed things up a bit. My GPU utilization does not go to 100% and keeps fluctuating. Maybe there is something wrong with my cpu or hard disk
